### PR TITLE
Root event date

### PR DESCRIPTION
**Pull request by thobson88**
Server-side code to support [trustchain-mobile #26](https://github.com/alan-turing-institute/trustchain-mobile/issues/26).

Adds an HTTP endpoint for getting candidate root DIDs (each with a corresponding Bitcoin transaction ID) that were timestamped on a given a calendar date.

Does **not** address the related #130.

### CODE REVIEW

File: trustchain-cli/src/bin/main.rs [hunk 1]
```diff
@@ -238,10 +239,11 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
                         let issuer = credential
                             .get_issuer()
                             .expect("No issuer present in credential.");
-                        let chain = TrustchainAPI::verify(issuer, root_event_time, &verifier)
-                            .await
-                            // Can unwrap as already verified above.
-                            .unwrap();
+                        let chain =
+                            TrustchainAPI::verify(issuer, root_event_time.into(), &verifier)
```
Comment (sgreenbury) [inline trustchain-cli/src/bin/main.rs:243]: Remove `into()` (clippy)

File: trustchain-core/src/utils.rs [hunk 2]
```diff
@@ -99,6 +100,15 @@ pub fn get_did_suffix(did: &str) -> &str {
     did.split(':').last().unwrap()
 }
 
+/// Converts a short-form DID into a complete DID.
+pub fn get_did_from_suffix(did_suffix: &str, method: &str) -> String {
+    let mut did = String::from_str("did:").unwrap();
+    did.push_str(method);
+    did.push_str(":");
+    did.push_str(did_suffix);
+    did
```
Comment (sgreenbury) [inline trustchain-core/src/utils.rs:109]:
```suggestion
format!("did:{method}:{did_suffix}")
```

File: trustchain-core/src/utils.rs [hunk 3]
```diff
@@ -99,6 +100,15 @@ pub fn get_did_suffix(did: &str) -> &str {
     did.split(':').last().unwrap()
 }
 
+/// Converts a short-form DID into a complete DID.
+pub fn get_did_from_suffix(did_suffix: &str, method: &str) -> String {
```
Comment (sgreenbury) [inline trustchain-core/src/utils.rs:104]: is method `ion:test` for testnet? Do we need an optional network instead to distinguish as in the `ION` type?

File: trustchain-http/src/errors.rs [hunk 4]
```diff
@@ -108,6 +118,7 @@ impl IntoResponse for TrustchainHTTPError {
             err @ TrustchainHTTPError::NoCredentialIssuer => {
                 (StatusCode::BAD_REQUEST, err.to_string())
             }
+            err @ TrustchainHTTPError::RootError(_) => (StatusCode::BAD_REQUEST, err.to_string()),
```
Comment (sgreenbury) [inline trustchain-http/src/errors.rs:121]: Check the HTTP status

File: trustchain-ion/src/root.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+/// Struct representing a root DID candidate.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct RootCandidate {
+    pub did: String,
+    pub txid: String,
+    pub block_height: u64,
+}
+
+/// Identifies potential root DIDs whose (UTC) timestamp matches a given date.
```
Comment (sgreenbury) [inline trustchain-ion/src/root.rs:56]: Line lengths

File: trustchain-ion/src/root.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+/// operation index zero (opIndex = 0). As such, root DIDs must be created in
+/// the first DID operation associated with a particular Bitcoin transaction.
+pub async fn root_did_candidates(
+    date: NaiveDate,
+) -> Result<Vec<RootCandidate>, TrustchainRootError> {
+    let block_height_range = block_height_range_on_date(date, None, None)?;
+    let cursor =
+        query_mongodb_on_interval(block_height_range.0 as u32, block_height_range.1 as u32).await?;
+
+    // The mongodb Cursor instance streams all bson documents that:
```
Comment (sgreenbury) [inline trustchain-ion/src/root.rs:67]: Remove?

File: trustchain-ion/src/verifier.rs [hunk 7]
```diff
@@ -134,20 +131,20 @@ where
 
     /// Gets RPC client.
     fn rpc_client(&self) -> &bitcoincore_rpc::Client {
-        self.rpc_client.as_ref().unwrap()
+        self.rpc_client.as_ref().unwrap() // TODO: handle failure here.
```
Comment (sgreenbury) [inline trustchain-ion/src/verifier.rs:134]: I think this is ok as the API has constrained it to exist here

File: trustchain-ion/src/verifier.rs [hunk 8]
```diff
@@ -134,20 +131,20 @@ where
...
     }
 
     /// Gets IPFS client.
     fn ipfs_client(&self) -> &IpfsClient {
-        self.ipfs_client.as_ref().unwrap()
+        self.ipfs_client.as_ref().unwrap() // TODO: handle failure here.
```
Comment (sgreenbury) [inline trustchain-ion/src/verifier.rs:139]: As above on 134

File: trustchain-ion/src/verifier.rs [hunk 9]
```diff
@@ -288,53 +285,54 @@ where
...
-
-        // Convert the block height to a block hash.
-        let block_hash = self
-            .rpc_client()
-            .get_block_hash(u64::from(block_height))
-            .map_err(|_| VerifierError::InvalidBlockHeight(block_height.into()))?;
-
-        Ok((block_hash, tx_index))
-    }
+    // MOVED TO utils.rs:
```
Comment (sgreenbury) [inline trustchain-ion/src/verifier.rs:288]: Remove?

File: trustchain-ion/src/root.rs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    #[tokio::test]
+    #[ignore = "Integration test requires Bitcoin & MongoDB"]
+    async fn test_root_did_candidates() {
+        let date = NaiveDate::from_ymd_opt(2022, 10, 20).unwrap();
+        let result = root_did_candidates(date).await.unwrap();
+
+        // There were 38 testnet ION operations with opIndex 0 on 20th Oct 2022.
+        // The block height range on that date is (2377360, 2377519).
+        // The relevant mongosh query is:
+        // db.operations.find({type: 'create', opIndex: 0, txnTime: { $gt: 2377359, $lt: 2377520}}).count()
```
Comment (sgreenbury) [inline trustchain-ion/src/root.rs:134]:
```suggestion
        //   `db.operations.find({type: 'create', opIndex: 0, txnTime: { $gt: 2377359, $lt: 2377520}}).count()`
```

File: trustchain-core/src/utils.rs [hunk 11]
```diff
@@ -99,6 +100,15 @@ pub fn get_did_suffix(did: &str) -> &str {
...
+    let mut did = String::from_str("did:").unwrap();
+    did.push_str(method);
+    did.push(':');
+    did.push_str(did_suffix);
+    did
```
Comment (sgreenbury) [inline trustchain-core/src/utils.rs:109]:
```suggestion
    format!("did:{method_and_network}:{did_suffix}")
```

File: trustchain-core/src/utils.rs (hunk 3 above)
Comment (sgreenbury) [inline trustchain-core/src/utils.rs:104]:
```suggestion
pub fn get_did_from_suffix(did_suffix: &str, method_and_network: &str) -> String {
```

File: trustchain-ion/src/root.rs (hunk 6 above)
Comment (thobson88) [reply trustchain-ion/src/root.rs:67]: Will leave for now as later refactoring will make obsolete.

**Review by edchapman88 — APPROVED** [review]
looks good!

**Pull request merged**